ocaml/tests/test_observer.ml

```diff
@@ -55,35 +144,218 @@ let start_test_trace () =
-        "Spans are collected in hashtable"
-        (Tracer.span_hashtbl_is_empty ())
-        false ;
-      Export.set_export_interval 0.2 ;
-      let _ = Tracer.finish x in
-      Unix.sleepf 0.2 ;
-      (* Wait for export to clear hashtbl *)
-      Alcotest.(check bool)
-        "Span export clears hashtable"
-        (Tracer.span_hashtbl_is_empty ())
-        true
-  | Error e ->
-      Alcotest.failf "Span start failed with %s" (Printexc.to_string e)
+let test_observer_create_and_destroy () =
+  let __context = Test_common.make_test_database () in
+  Xapi_DB.assert_num_observers ~__context 0 ;
+  TracerProvider.assert_num_observers ~__context 0 ;
 
-let raise_exn () = raise (Failure "Test exception message")
+  let name = "test-observer" in
+  let self = test_create ~__context () ~name_label:name in
+  Xapi_DB.assert_num_observers ~__context 1 ;
+  TracerProvider.assert_num_observers ~__context 1 ;
 
-let test_b () = raise_exn () + 1 (* non-tail to ensure stack entry created *)
+  Xapi_DB.assert_observer_disabled ~__context ~self ;
+  TracerProvider.assert_observer_disabled ~name ;
 
-let test_a () = test_b () + 1
+  (* DB has no mandatory attributes *)
+  TracerProvider.assert_mandatory_attributes ~name ;
+
+  let name2 = "test-observer-2" in
+  let self2 = test_create ~__context ~name_label:name2 () in
+  Xapi_DB.assert_num_observers ~__context 2 ;
+  TracerProvider.assert_num_observers ~__context 2 ;
+
+  test_destroy ~__context ~self:self2 () ;
+  Xapi_DB.assert_num_observers ~__context 1 ;
+  TracerProvider.assert_num_observers ~__context 1 ;
+
+  test_destroy ~__context ~self () ;
+  Xapi_DB.assert_num_observers ~__context 0 ;
+  TracerProvider.assert_num_observers ~__context 0
+
+let test_hosts ~__context ~self =
+  let valid_host = !Xapi_globs.localhost_ref in
+  ( try Xapi_observer.set_hosts ~__context ~self ~value:[valid_host]
+    with _ -> Alcotest.failf "Failed to set valid host"
+  ) ;
+
+  let invalid_host = Ref.null in
+  Alcotest.check_raises "Xapi_observer.set_hosts should fail on invalid host"
+    Api_errors.(
+      Server_error (invalid_value, ["host"; Ref.string_of invalid_host])
+    )
+    (fun () ->
+      Xapi_observer.set_hosts ~__context ~self ~value:[invalid_host] |> ignore
+    )
+
+let test_components ~__context ~self =
+  let valid_components = ["xapi"] in
+  List.iter
+    (fun x ->
+      try Xapi_observer.set_components ~__context ~self ~value:[x]
+      with _ -> Alcotest.failf "Failed to set valid component"
+    )
+    valid_components ;
+
+  let invalid_component = "invalid-component" in
+  Alcotest.check_raises
+    "Xapi_observer.set_components should fail on invalid component"
+    Api_errors.(Server_error (invalid_value, ["component"; invalid_component]))
+    (fun () ->
+      Xapi_observer.set_components ~__context ~self ~value:[invalid_component]
+      |> ignore
+    )
+
+let test_endpoints ~__context ~self =
+  let valid_endpoints = ["bugtool"; "http://example.com:9411/api/v2/spans"] in
```

Comment:
It would be interesting to add at least an invalid endpint, it could be used to accumulate regression tests for invalid endpoints

Reply:
because currently the only endpoint validity check we do is Uri.to_url, I believe it accepts _any_ string so the test would always fail -Can always add an invalid_endpoint check when/if we change this

Reply:
The validation happens in ocaml/xapi/xapi_observer.ml -> validate_endpoint. Uri.of_string is used to validate the string, on top of that the scheme is enforced and the host of the uri is also validated to be a valid hostname or an IP. I think there can be lots of invalid strings, and even invalid URIs as well :)

Reply:
oh whoops! that's new, I thought we were adding proper validation later -my bad